File: dice.py
import random

class Dice:
    def __init__(self):
        self._face_value = random.randint(1, 6)
    
    @property
    def face_value(self):
        return self._face_value
    
    @face_value.setter
    def face_value(self, value):
        if 1 <= value <= 6:
            self._face_value = value
        else:
            raise ValueError("Face value must be between 1 and 6")
    
    def roll(self):
        self._face_value = random.randint(1, 6)
        return self._face_value

File: yahtzee_scala/src/main/scala/yahtzee/Dice.scala
package yahtzee

opaque type Dice = Int

object Dice:
  def apply(value: Int): Dice = 
    require(value >= 1 && value <= 6, "Face value must be between 1 and 6")
    value
  
  extension (d: Dice)
    def value: Int = d
    
    def roll(seed: Long): (Dice, Long) =
      val newSeed = (seed * 0x5DEECE66DL + 0xBL) & 0xFFFFFFFFFFFFL
      val newValue = ((newSeed >>> 16).toInt.abs % 6) + 1
      (Dice(newValue), newSeed)
  
  def create(seed: Long): (Dice, Long) =
    val newSeed = (seed * 0x5DEECE66DL + 0xBL) & 0xFFFFFFFFFFFFL
    val value = ((newSeed >>> 16).toInt.abs % 6) + 1
    (Dice(value), newSeed)
